File: src/numbl-core/ops/realUnaryElemwise.ts
/**
 * Pure-TS implementation of real unary element-wise ops.
 * Mirrors native/ops/real_unary_elemwise.c.
 */

import { OpUnary } from "./opCodes.js";

function rsign(x: number): number {
  return x > 0 ? 1 : x < 0 ? -1 : 0;
}

export function tsRealUnaryElemwise(
  op: number,
  n: number,
  a: Float64Array,
  out: Float64Array
): void {
  switch (op) {
    case OpUnary.EXP:
      for (let i = 0; i < n; i++) out[i] = Math.exp(a[i]);
      return;
    case OpUnary.LOG:
      for (let i = 0; i < n; i++) out[i] = Math.log(a[i]);
      return;
    case OpUnary.LOG2:
      for (let i = 0; i < n; i++) out[i] = Math.log2(a[i]);
      return;
    case OpUnary.LOG10:
      for (let i = 0; i < n; i++) out[i] = Math.log10(a[i]);
      return;
    case OpUnary.SQRT:
      for (let i = 0; i < n; i++) out[i] = Math.sqrt(a[i]);
      return;
    case OpUnary.ABS:
      for (let i = 0; i < n; i++) out[i] = Math.abs(a[i]);
      return;
    case OpUnary.FLOOR:
      for (let i = 0; i < n; i++) out[i] = Math.floor(a[i]);
      return;
    case OpUnary.CEIL:
      for (let i = 0; i < n; i++) out[i] = Math.ceil(a[i]);
      return;
    case OpUnary.ROUND:
      // MATLAB round: half-away-from-zero. Math.round is half-up (towards +Inf),
      // which differs for negative halves (e.g. Math.round(-0.5)=0 vs MATLAB -1).
      for (let i = 0; i < n; i++) {
        const x = a[i];
        out[i] = x >= 0 ? Math.round(x) : -Math.round(-x);
      }
      return;
    case OpUnary.TRUNC:
      for (let i = 0; i < n; i++) out[i] = Math.trunc(a[i]);
      return;
    case OpUnary.SIN:
      for (let i = 0; i < n; i++) out[i] = Math.sin(a[i]);
      return;
    case OpUnary.COS:
      for (let i = 0; i < n; i++) out[i] = Math.cos(a[i]);
      return;
    case OpUnary.TAN:
      for (let i = 0; i < n; i++) out[i] = Math.tan(a[i]);
      return;
    case OpUnary.ASIN:
      for (let i = 0; i < n; i++) out[i] = Math.asin(a[i]);
      return;
    case OpUnary.ACOS:
      for (let i = 0; i < n; i++) out[i] = Math.acos(a[i]);
      return;
    case OpUnary.ATAN:
      for (let i = 0; i < n; i++) out[i] = Math.atan(a[i]);
      return;
    case OpUnary.SINH:
      for (let i = 0; i < n; i++) out[i] = Math.sinh(a[i]);
      return;
    case OpUnary.COSH:
      for (let i = 0; i < n; i++) out[i] = Math.cosh(a[i]);
      return;
    case OpUnary.TANH:
      for (let i = 0; i < n; i++) out[i] = Math.tanh(a[i]);
      return;
    case OpUnary.SIGN:
      for (let i = 0; i < n; i++) out[i] = rsign(a[i]);
      return;
    default:
      throw new Error(`tsRealUnaryElemwise: unknown op ${op}`);
  }
}

File: native/ops/real_unary_elemwise.c
/**
 * Real unary element-wise ops (op-code dispatch).
 * Caller-allocated input/output buffers; never copies.
 */

#include "numbl_ops.h"

#include <math.h>

static inline double rsign(double x) {
  return x > 0.0 ? 1.0 : x < 0.0 ? -1.0 : 0.0;
}

int numbl_real_unary_elemwise(int op, size_t n,
                              const double* a, double* out) {
  if (!a || !out) return NUMBL_ERR_NULL_PTR;
  switch (op) {
    case NUMBL_UNARY_EXP:
      for (size_t i = 0; i < n; i++) out[i] = exp(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_LOG:
      for (size_t i = 0; i < n; i++) out[i] = log(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_LOG2:
      for (size_t i = 0; i < n; i++) out[i] = log2(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_LOG10:
      for (size_t i = 0; i < n; i++) out[i] = log10(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_SQRT:
      for (size_t i = 0; i < n; i++) out[i] = sqrt(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_ABS:
      for (size_t i = 0; i < n; i++) out[i] = fabs(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_FLOOR:
      for (size_t i = 0; i < n; i++) out[i] = floor(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_CEIL:
      for (size_t i = 0; i < n; i++) out[i] = ceil(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_ROUND:
      for (size_t i = 0; i < n; i++) out[i] = round(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_TRUNC:
      for (size_t i = 0; i < n; i++) out[i] = trunc(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_SIN:
      for (size_t i = 0; i < n; i++) out[i] = sin(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_COS:
      for (size_t i = 0; i < n; i++) out[i] = cos(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_TAN:
      for (size_t i = 0; i < n; i++) out[i] = tan(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_ASIN:
      for (size_t i = 0; i < n; i++) out[i] = asin(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_ACOS:
      for (size_t i = 0; i < n; i++) out[i] = acos(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_ATAN:
      for (size_t i = 0; i < n; i++) out[i] = atan(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_SINH:
      for (size_t i = 0; i < n; i++) out[i] = sinh(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_COSH:
      for (size_t i = 0; i < n; i++) out[i] = cosh(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_TANH:
      for (size_t i = 0; i < n; i++) out[i] = tanh(a[i]);
      return NUMBL_OK;
    case NUMBL_UNARY_SIGN:
      for (size_t i = 0; i < n; i++) out[i] = rsign(a[i]);
      return NUMBL_OK;
    default:
      return NUMBL_ERR_BAD_OP;
  }
}
